extensions/amp-story/1.0/amp-story-page.js

```diff
@@ -322,20 +327,163 @@ export class AmpStoryPage extends AMP.BaseElement {
     this.maybeConvertCtaLayerToPageOutlink_();
   }
 
-  /** @private */
-  maybeSetPreviewDuration_() {
-    if (this.storeService_.get(StateProperty.PREVIEW_STATE)) {
-      const videos = this.getAllVideos_();
+  /**
+   * Handles visibility state changes.
+   * @private
+   */
+  onVisibilityChanged_() {
+    const ampDoc = this.getAmpDoc();
 
-      const autoAdvanceAttr =
-        videos.length > 0
-          ? VIDEO_PREVIEW_AUTO_ADVANCE_DURATION
-          : DEFAULT_PREVIEW_AUTO_ADVANCE_DURATION;
+    const wasPreview = this.visibilityState_ === VisibilityState_Enum.PREVIEW;
+    const isPreviewToVisibleTransition = wasPreview && ampDoc.isVisible();
+    this.visibilityState_ = ampDoc.getVisibilityState();
 
-      this.element.setAttribute('auto-advance-after', autoAdvanceAttr);
+    if (ampDoc.isPreview() || ampDoc.isVisible()) {
+      // Preview mode and visible mode both have different advancement logic.
+      this.setUpAdvancementConfig_(isPreviewToVisibleTransition);
     }
   }
 
+  /**
+   * Sets up the advancement config depending upon the
+   * @param {boolean=} handlePreviewToVisibleTransition
+   * @private
+   */
+  setUpAdvancementConfig_(handlePreviewToVisibleTransition = false) {
+    if (this.getAmpDoc().isPreview()) {
+      this.setupAutoAdvanceForPreview_();
+      this.initializeAdvancementConfig_();
+    }
+
+    if (this.getAmpDoc().isVisible()) {
+      this.setupAutoAdvanceForVisible_();
+      this.maybeSetStoryNextUp_();
+      if (this.isActive() && handlePreviewToVisibleTransition) {
+        this.handlePreviewToVisibleTransition_();
+      } else {
+        this.initializeAdvancementConfig_();
+      }
+    }
+  }
+
+  /**
+   * Configures the page to auto advance using preview-specific durations.
+   * @private
+   */
+  setupAutoAdvanceForPreview_() {
+    let previewSecondsPerPage = this.viewer_.getParam('previewSecondsPerPage');
+    if (previewSecondsPerPage) {
+      previewSecondsPerPage = parseInt(previewSecondsPerPage, 10);
+    }
```

Comment:
parseInt will return NaN if you parse `null` so it might be safe to include the `parseInt` in the first line: `const previewSecondsPerPage = parseInt(this.viewer_.getParam('previewSecondsPerPage'), 10)`